### DISCUSSION TITLE
ID van observation goal, sensor, en datastroom achterhalen

### DISCUSSION THREAD

**Discussion by CelineJansen** [kadaster-labs/sensrnet-home#262] [Q&A] [answered]
In de UI van de demo-omgeving kan ik IDs ophalen van apparaten / devices. Voor observation goal, sensor, en datastroom zie ik de IDs (nog) niet 1-2-3 terug in de UI. Dat is best onhandig als ik in bulk aanpassingen wil doen met behulp van de API (ik was vandaag bulk verwijderingen aan het testen). 

Wanneer ik via de API apparaten, sensoren, datastromen, observatie doelen opvoer krijg ik als response (als alles goed gaat) een ID. Dit ID kan ik natuurlijk wel opslaan en voor latere bewerkingen / verwijderingen met de API benutten. Desondanks zou de optie om e.e.a. te controleren in de UI van toegevoegde waarde zijn. Voor het probleem waar ik nu tegenaan loop is dit zelfs noodzakelijk:

Een tijdje terug heb ik in bulk apparaten, sensoren, en datastromen opgevoerd. Het ID van de sensoren dat ik destijds als response kreeg heb ik opgeslagen. Ik wil nu proberen de device te behouden, en alleen de sensor te verwijderen (DELETE request: https://demo.sensorenregister.nl/api/device/@Value(deviceId)/sensor/@Value(sensorId)). Dit lukt echter niet met het ID uit de response dat ik heb opgeslagen. Dit is de response die ik dan krijg: 

![image](https://user-images.githubusercontent.com/79267472/141788092-8c401f41-8486-4504-b5c6-3eaba0d6ced8.png)

(Het in bulk verwijderen van apparaten op basis van het eerder verkregen ID werkt overigens wel.)

Mijn vraag is dus tweedelig:
- Hebben jullie een idee waarom ik bovenstaande foutmelding krijg? Kan het ID uit de response van destijds tussentijds gewijzigd zijn of heb ik gewoon een fout gemaakt met opslaan? (Als ik als test nu een nieuwe sensor opvoer, en dat ID uit de nieuwe response gebruik werkt het wel, maar ik wil dus graag weten of het voor langere termijn opslaan van IDs voor hergebruik op een later moment een werkbare oplossing is, of dat deze IDs nog kunnen wijzigen en opslaan geen zin heeft).
- Kan worden ingebouwd dat de IDs van sensoren, datastromen, en observatie doelen in de UI zijn terug te vinden (dat is nodig om fouten na te lopen, ook in productie na de pilot lijkt mij).

**Comment by marcvanandel** [comment]
Hele goeie vraag!! Ik heb veel te weinig aandacht gegeven aan LinkedData ... maar ik had dat graag nog toe willen voegen aan het MVP. En met 'dat' bedoel ik semantiek, onthologie en informatie model. We hebben wel een informatie model ... in de vorm van een datamodel. Verdere LinkedData en semantiek hebben we (nog) niet uitgewerkt.

LinkedData heeft per definitie duidelijke IDs - in de vorm van URIs. Lijkt mij een hele goede uitbreiding ... maar daar heb je nu niks aan 😏

**Comment by kad-busses** [comment, marked as answer]
- De deviceID in de screenshots hoort bij device "Geluidssensor 304", terwijl de sensorId bij de sensor van device "Geluidssensor 306" hoort. Daardoor kan de API de sensor niet vinden. Met de juiste deviceId/sensorId zou het gewoon moeten werken.

- Om de IDs te achterhalen kan op dit moment de API gebruikt worden, daar komen de IDs gewoon uit terug, al zijn ze een beetje verstopt. De sensor- en datastroominformatie wordt namelijk onder het device opgeslagen. Wanneer een device een sensor en datastream bevat, kun je de informatie daar dus terug vinden. Met een GET verzoek op `/api/device/{deviceId}` komen ook de velden `sensors` en `dataStreams` terug, beiden als array. Elke entry heeft een `_id` veld. Elke datastream heeft daarnaast een veld `observationGoalIds`, waar je de observatiedoelen Ids terug kunt vinden. Met GET `/api/device` (op basis van legalEntityId) komt er overigens ook een volledige lijst terug met alle devices, inclusief sensoren, datastromen en observation goals. Deze informatie zou voldoende moeten zijn om de bulkbewerkingen programmatisch te kunnen doen 🙂

**Reply by CelineJansen** [reply]
Super goede tip. Dan komen die IDs in mijn script blijkbaar niet 'chronologisch' terug. Maar dat is mijn probleem en kan ik nu tenminste controleren. Met die GET request via api/device kan ik mooi een 0-stand opvragen. Al is het dan (nu nog) een beetje verstopt, fijn dat het tenminste kan.

**Reply by marcvanandel** [reply]
Gelukkig antwoord @kad-busses wél op de vraag 😁 en wordt je niet alleen maar met een kluitje 'toekomstbeeld' in het riet gestuurd 🙈
